keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1149 @@
+and make progress.
+-->
+
+- Replace existing usage of Protobuf.
+- Substantially reduce the size of encoded objects (a modest size reduction is
+  anticipated).
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Format
+
+The output of the CBOR encoder is a single tagged data item as specified in
+“[Self-Described
+CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-self-described-cbor)”,
+with no additional envelope. Self-described CBOR -- a tagged data item with tag
+number 55799 -- has the same semantics as the same data item with no tag, with
+the convenient property that its encoded form is always prefixed by
+`0xd9d9f7`. By design, this prefix is never found at the beginning of a JSON
+text and can be used as a “magic number” to distinguish the data format of a
+stored object at rest.
+
+To support decoding custom resources that have been stored as a mixture of JSON
+and CBOR, the CBOR serializer will implement RecognizingDecoder by checking for
+the prefix `0xd9d9f7`.
+
+Streaming responses (i.e. watches) will be serialized as CBOR Sequences. A [CBOR
+Sequence](https://www.rfc-editor.org/rfc/rfc8742.html) is a concatenation of
+zero or more CBOR data items, with no additional framing. This is effectively
+equivalent to the existing JSON stream serialization behavior and takes
+advantage of the property that, like JSON objects – and unlike non-object JSON
+documents, e.g. numbers – CBOR data items are self-delimiting.
+
+### Negotiation
+
+Proactive content negotiation will be supported for clients that want to receive
+CBOR-encoded responses using the MIME type “application/cbor” in the Accept
+request header. For compatibility with API servers that don’t support CBOR,
+clients should also accept “application/json” (with a lower quality factor) and
+choose the appropriate decoder based on the Content-Type response header.
+
+Similarly, streaming requests should use the MIME type for CBOR Sequences,
+“application/cbor-seq”.
+
+CBOR will not be a supported encoding for JSON Patches, JSON Merge Patches,
+Strategic Merge Patches, or Server-Side Apply configurations.
+
+Clients can send CBOR-encoded request bodies with the appropriate Content-Type
+to API servers that support CBOR. API servers that don’t support CBOR will
+return status 415. Rather than implementing a mechanism up-front to negotiate
+CBOR-encoded requests from clients (for example: optimistic CBOR with fallback
+to JSON on 415, or advertising supported media types and checking them in
+advance), dynamic clients will continue to default to JSON for writes.
```

Comment:
Forever?

Even if JSON is the default, shouldn't we define how we'd like clients to handle fallback?  Or at least define how we'll have client-go handle it?

Reply:
Elaborated on both issues, thanks for pointing this out.